### PR TITLE
Add rate limiter priority to ReadOptions

### PR DESCRIPTION
**Pull request by ajkr**
Users can set the priority for file reads associated with their operation by setting `ReadOptions::rate_limiter_priority` to something other than `Env::IO_TOTAL`. Rate limiting `VerifyChecksum()` and `VerifyFileChecksums()` is the motivation for this PR, so it also includes benchmarks and minor bug fixes to get that working.

`RandomAccessFileReader::Read()` already had support for rate limiting compaction reads. I changed that rate limiting to be non-specific to compaction, but rather performed according to the passed in `Env::IOPriority`. Now the compaction read rate limiting is supported by setting `rate_limiter_priority = Env::IO_LOW` on its `ReadOptions`.

There is no default value for the new `Env::IOPriority` parameter to `RandomAccessFileReader::Read()`. That means this PR goes through all callers (in some cases multiple layers up the call stack) to find a `ReadOptions` to provide the priority. There are TODOs for cases I believe it would be good to let user control the priority some day (e.g., file footer reads), and no TODO in cases I believe it doesn't matter (e.g., trace file reads).

The API doc only lists the missing cases where a file read associated with a provided `ReadOptions` cannot be rate limited. For cases like file ingestion checksum calculation, there is no API to provide `ReadOptions` or `Env::IOPriority`, so I didn't count that as missing.

Test Plan:
- new unit tests
- new benchmarks on ~50MB database with 1MB/s read rate limit and 100ms refill interval; verified with strace reads are chunked (at 0.1MB per chunk) and spaced roughly 100ms apart.
  - setup command: `./db_bench -benchmarks=fillrandom,compact -db=/tmp/testdb -target_file_size_base=1048576 -disable_auto_compactions=true -file_checksum=true`
  - benchmarks command: `strace -ttfe pread64 ./db_bench -benchmarks=verifychecksum,verifyfilechecksums -use_existing_db=true -db=/tmp/testdb -rate_limiter_bytes_per_sec=1048576 -rate_limit_bg_reads=1 -rate_limit_user_ops=true -file_checksum=true`
- crash test using IO_USER priority on non-validation reads with #9567 reverted: `python3 tools/db_crashtest.py blackbox --max_key=1000000 --write_buffer_size=524288 --target_file_size_base=524288 --level_compaction_dynamic_level_bytes=true --duration=3600 --rate_limit_bg_reads=true --rate_limit_user_ops=true --rate_limiter_bytes_per_sec=10485760 --interval=10`

### CODE REVIEW

File: db/db_test2.cc
Comment (hx235) [inline db/db_test2.cc:3980]: nit: you may want to `ASSERT_EQ(options.rate_limiter->GetTotalBytesThrough(Env::IO_LOW), options.rate_limiter->GetTotalBytesThrough(Env::IO_TOTAL));` to make sure compaction read was charged at ENV::IO_LOW only.

File: include/rocksdb/options.h
Comment (hx235) [inline include/rocksdb/options.h:1603]: "..with this option.." instead of operation?

File: include/rocksdb/options.h
Comment (hx235) [inline include/rocksdb/options.h:1613]: I am thinking maybe a more user-relevant way of stating this can be:

```
List of lower-level reads that support rate-limiting and take ReaderOptions::priority into account include: X, Y, Z (list is subjected to future expansion). 
For others, there isn't any rate-limiting and ReaderOptions::priority will either be ignored (e.g, plain/cuckoo table reads, old BlobDB reads, and generally small reads like or file headers/footers) or requires ReaderOptions::priority to be Env::IO_TOTAL (e.g, RandomAccessFile::MultiRead()/DB::MultiGet()) .
```

File: include/rocksdb/options.h
Comment (hx235) [inline include/rocksdb/options.h:1616]: Is there a reason why we go for a general name `priority`, instead of something like `rate_limiter_priority` which might better avoid future collision with priority related field being added here?

**Review by hx235 — COMMENTED** [review]
Early review: I looked at the code change/testing for compaction and VerifyChecksum and believe it's sufficient for delivery with a few nits. 

Will look other read-related changes in the second round. If there is a way to split them into two PRs, I am good with approving the compaction/Verifychecksum related one. 

One thing though is we might want to work on a better API doc for both DBOptions::rate_limiter and ReadOptions::priority since (1) we expand the scope of things being rate limited by DBOptions::rate_limiter (2) we open a way to bypass that rate limiting using ReadOptions::priority (3) some lower-level operation will ignore ReadOptions::priority no matter what.

I'm thinking of the following comment:
(1)For `DBOptions::rate_limiter` API:
If not null, bytes_per_sync is set to 1MB by default and it will be used to rate-limit the following (the list is subjected to future expansion):
Read:
- Compaction read (at IOPriority::LOW)
- DB::VerifyChecksum(at specified priority or be bypassed, see ReadOptions::priority for more)
Write:
- Compaction write (at IOPriority::LOW)
- Flush write (at IOPriority::HIGH)

Default: nullptr (rate limiting disabled)

(2)ReadOptions::rate_limiter: see PR comment

**Review by ajkr — COMMENTED** [review]
Thanks for the review!

> Early review: I looked at the code change/testing for compaction and VerifyChecksum and believe it's sufficient for delivery 👍 with a few nits.
> 
> Will look at other read-related changes in the second round. If there is a way to split them into two PRs, I am good with approving the compaction/Verifychecksum related one.

Sorry the main idea of this PR requires them to be quite entangled (see below).

> 
> One thing though is we might want to work on a better API doc for both DBOptions::rate_limiter and ReadOptions::priority since (1) we expand the scope of things being rate limited by DBOptions::rate_limiter (2) we open a way to bypass that rate limiting using ReadOptions::priority (3) some lower-level operation will ignore ReadOptions::priority no matter what.
> 
> I'm thinking of the following comment: (a) For `DBOptions::rate_limiter` API:
> 
> ```
> If not null, bytes_per_sync is set to 1MB by default and it will be used to rate-limit the following (the list is subjected to future expansion):
> Read:
> - Compaction read (at IOPriority::LOW)
> - DB::VerifyChecksum(at specified priority or be bypassed, see ReadOptions::priority for more)
> Write:
> - Compaction write (at IOPriority::LOW)
> - Flush write (at IOPriority::HIGH)
> 
> Default: nullptr (rate limiting disabled)
> ```
> 
> (b) ReadOptions::rate_limiter: see PR comment

The goal of this PR is to generally support `ReadOptions::priority` besides in  unimportant cases and a handful of uncommon cases.

It is easier to describe where support is missing than where support is available. Describing either one should be acceptable, IMO, since it's the complement of the other. I am happy to reexamine the list of missing support, both for making sure it is at the proper level of abstraction (for example, the reference to `MultiRead()` is too low-level) and making sure it is complete (for example, I have not checked for APIs accepting `ReadOptions`  that do their reads outside `RandomAccessFileReader`).

It is difficult to list available support because this PR made changes at a low level to support most user APIs (without needing to look at what exactly they are) at once. For example, a lot of APIs (`Get()`, iterator, compaction, etc.) already passed their `ReadOptions` all the way down to `BlockFetcher`. I only made a small code change to `BlockFetcher` to make rate limiting available to all those APIs. That is also why this approach requires we do it all at once -- if we wanted to only support compaction/VerifyChecksum, we'd have to go into `BlockFetcher` and explicitly remove support for all other cases, which seems like extra code for an unnecessary limitation.

Meanwhile, it was actually fairly easy to identify the user API corresponding to missing support. I guess it's because that case is (luckily) much less common, where the user can provide `ReadOptions` that are inaccessible to `RandomAccessFileReader`.

File: db/db_test2.cc
Comment (ajkr) [reply db/db_test2.cc:3980]: Sure, will do.

**Comment by hx235** [comment]
Aha - I could see that now. Read operations that pass `ReadOptions` down to `RandomAccessFileReader` can now choose to have rate limiting support, the list of which is pretty big compared with the counterpart. The list will be further expanded once we cover SequentialFileReader. 

> I am happy to reexamine the list of missing support, both for making sure it is at the proper level of abstraction (for example, the reference to MultiRead() is too low-level) and making sure it is complete (for example, I have not checked for APIs accepting ReadOptions that do their reads outside RandomAccessFileReader).

This will be great. I'm good with listing unsupported user read API (such as DB::MultiGet(), old BlobDB reads, general reads under plain/cuckoo table configuration) in `DBOptions::rate_limiter` API doc so that users know what NOT to expect for rate limiting support. It will also be good to briefly mention there something like "see `ReadOptions::priority` for more info on specifying rate-limiting priority of each read or bypassing the rate limiting" so that users know how correctly add such rate-limiting support. For internal read/write of compaction and write of flush, contrary to user read API, their priorities are fixed and will not be bypassed using `ReadOptions::priority`, which is basically what the existing `DBOptions::rate_limiter` API describes.

I will also update the comment as getting SequentialFileReader and write-side of things done.

File: include/rocksdb/options.h
Comment (ajkr) [reply include/rocksdb/options.h]: Done.

**Review by ajkr — COMMENTED** [review]
> This will be great. I'm good with listing unsupported user read API (such as DB::MultiGet(), old BlobDB reads, general reads under plain/cuckoo table configuration) in DBOptions::rate_limiter API doc so that users know what NOT to expect for rate limiting support. It will also be good to briefly mention there something like "see ReadOptions::priority for more info on specifying rate-limiting priority of each read or bypassing the rate limiting" so that users know how correctly add such rate-limiting support.

I put the list in one place (`ReadOptions::priority`) and referred to it from the `DBOptions::rate_limiter`. The list of scenarios that ignore/do not support `ReadOptions::priority` is shorter than the list of scenarios where read rate limiting is unavailable. Scenarios like reading MANIFEST/WALs during `DB::Open()` do not support read rate limiting, but I do not think we need to explicitly list cases where the API offers no way to enable rate limiting (there is no `ReadOptions` passed to `DB::Open()`, for example).

Also done:
- Went through all clients of `SequentialFileReader` and did not find any where the user-facing API offers a way to enable rate limiting.
- Updated to specify which `MultiGet()` APIs are supported at the proper level of abstraction.
- Removed mention of old `BlobDB` since it was never actually exposed in any public header.
- Elaborated on "cuckoo/plain table" so people can identify whether they are relevant to their use case

> For internal read/write of compaction and write of flush, contrary to user read API, their priorities are fixed and will not be bypassed using ReadOptions::priority, which is basically what the existing DBOptions::rate_limiter API describes.

Right. Keep in mind there is no way for a user to supply the `ReadOptions` for flush or compaction, so we don't really have a choice about bypassing `ReadOptions::priority` or not.

**Comment by hx235** [comment]
> The list of scenarios that ignore/do not support ReadOptions::priority is shorter than the list of scenarios where read rate limiting is unavailable. Scenarios like reading MANIFEST/WALs during DB::Open() do not support read rate limiting, but I do not think we need to explicitly list cases where the API offers no way to enable rate limiting (there is no ReadOptions passed to DB::Open(), for example).

Sounds good - thanks for clarification. Sounds like now we make ReadOptions a necessary condition to having rate limiting support for user read. Then we either take ReadOptions::priority into account (i.e, user specifying a priority or RocksDB internal sets IO_TOTAL as not-supported) or ignore it.

**Comment by hx235** [comment]
Not a rush to finishing the PR but are we still going to change the name for ReadOptions::priority? Since it's a public API, it will be good finalize the naming as early as possible.

File: include/rocksdb/options.h
Comment (ajkr) [reply include/rocksdb/options.h]: Done.

**Review by ajkr — COMMENTED** [review]
> Not a rush to finishing the PR but are we still going to change the name for ReadOptions::priority? Since it's a public API, it will be good to finalize the naming as early as possible.

Yes, it is done now. Thanks again for the detailed review!

**Comment by hx235** [comment]
Will get back to the second review very soon.

File: db/db_rate_limiter_test.cc
Comment (hx235) [inline db/db_rate_limiter_test.cc:212]: nit: might worth a comment here explaining why we branches for `use_block_cache_` or not in backward iteration but not forward

File: db/db_rate_limiter_test.cc
Comment (hx235) [inline db/db_rate_limiter_test.cc:217]: Good comment!

File: db/db_test2.cc
Comment (hx235) [inline db/db_test2.cc:3986]: I found it interesting that `direct_io_extra` of  512KB data compaction input can be up to equally large as 512KB.

File: file/file_util.h
Comment (hx235) [inline file/file_util.h:56]: Is removing this inline function one of those minor fixes that goes along with this PR?

**Comment by hx235** [comment]
Second round review: 
- unit tests seems sufficient and very clearly written 👍
- left 1 nit and 1 question
- might worth adding an option `FLAG_read_rate_limiter_priority` to stress test considering it touches many aspects of our read. Hopefully it shouldn't be hard as a quick search in `db_stress_test_base` reveals 11 `ReadOptions`.

File: db/db_test2.cc
Comment (ajkr) [reply db/db_test2.cc:3986]: TBH I just changed the indentation here; I have not read this lately but will trust you it's interesting :).

**Review by ajkr — COMMENTED** [review]
Thanks for the review! Replied for now and will make code updates later.

File: db_stress_tool/db_stress_test_base.cc
Comment (hx235) [inline db_stress_tool/db_stress_test_base.cc:352]: 👍

**Review by hx235 — APPROVED** [review]
Do we also need to add it to db_crashtest.py for automating the test?

Others LGTM!

**Review by ajkr — COMMENTED** [review]
> might worth adding an option FLAG_read_rate_limiter_priority to stress test considering it touches many aspects of our read. Hopefully it wouldn’t be hard as a quick search in db_stress_test_base reveals 11 ReadOptions.

Done. Although, keep in mind the crash test wrapper does not use rate limiter yet, so the support is a bit superficial for now. We will enable rate limiting eventually, but not in the near future (`IngestExternalFile()` is my next step to enable). I at least ran it manually and added the command to the test plan to show the new feature can work.

**Comment by ajkr** [comment]
> Do we also need to add it to db_crashtest.py for automating the test?

Sorry I wrote my last comment without refreshing. Basically yes, but it's non-trivial to enable it without slowing things down too much, so I hope to defer it.

**Comment by hx235** [comment]
> > Do we also need to add it to db_crashtest.py for automating the test?
> 
> Sorry I wrote my last comment without refreshing. Basically yes, but it's non-trivial to enable it without slowing things down too much, so I hope to defer it.

Sure! I was actually just referring to something like `"rate_limit_user_ops": lambda: random.choice([0, 1])` cuz I didn't realize "crash test wrapper does not use rate limiter yet" :p

**Comment by hx235** [comment]
> edit: Another reason I'm seeing it's nontrivial as it runs is it's printing a lot of these lines:

Got you! Might be good to just leave a comment `//"rate_limit_user_ops": lambda: random.choice([0, 1])` in case we (or I) forget to add it later.

**Comment by ajkr** [comment]
> > edit: Another reason I'm seeing it's nontrivial as it runs is it's printing a lot of these lines:
> 
> Got you! Might be good to just leave a comment `//"rate_limit_user_ops": lambda: random.choice([0, 1])` in case we (or I) forget to add it later.

Not sure, I think the person to enable rate limiting in CI will need to do a full tour of the options. There's a bunch of things I would do if we did it right now, like renaming rate_limit_bg_reads, making it support kAllIo (currently we only support reads or writes, never both), picking a rate_limiter_bytes_per_sec that doesn't slow things down, dynamically changing the rate limit, enabling IO_USER, etc. etc. But I guess whoever eventually does it will have to explore a bit to figure out the right way. I kind of don't think they'll miss the flags with "rate_limit" in the name :).

**Comment by hx235** [comment]
> > > edit: Another reason I'm seeing it's nontrivial as it runs is it's printing a lot of these lines:
> > 
> > 
> > Got you! Might be good to just leave a comment `//"rate_limit_user_ops": lambda: random.choice([0, 1])` in case we (or I) forget to add it later.
> 
> Not sure, I think the person to enable rate limiting in CI will need to do a tour of all the possibilities. There's a bunch of things I would do if we did it right now, like renaming rate_limit_bg_reads, making it support kAllIo (currently we only support reads or writes, never both), picking a rate_limiter_bytes_per_sec that doesn't slow things down, dynamically changing the rate limit, enabling IO_USER, etc. etc. But I guess whoever eventually does it will have to explore a bit to figure out the right way. I kind of don't think they'll miss the flags with "rate_limit" in the name :).

Sounds good! I would likely check back on this comment if I happen to be that person to enable rate limiting in CI .....

**Pull request closed without merging**
